Mathlib/NumberTheory/ModularForms/NormTrace.lean

```diff
@@ -0,0 +1,271 @@
+/-
+Copyright (c) 2025 David Loeffler. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: David Loeffler
+-/
+module
+
+public import Mathlib.NumberTheory.ModularForms.LevelOne
+
+/-!
+# Norm and trace maps
+
+Given two subgroups `𝒢, ℋ` of `GL(2, ℝ)` with `𝒢.relindex ℋ ≠ 0` (i.e. `𝒢 ⊓ ℋ` has finite index
+in `ℋ`), we define a trace map from `ModularForm (𝒢 ⊓ ℋ) k` to `ModularForm ℋ k`.
+-/
+@[expose] public noncomputable section
+
+open UpperHalfPlane
+
+open scoped ModularForm Topology Filter Manifold
+
+variable {𝒢 ℋ : Subgroup (GL (Fin 2) ℝ)}
+  {F : Type*} (f : F) [FunLike F ℍ ℂ] {k : ℤ}
+
+lemma IsCusp.mono {c : OnePoint ℝ} (hGH : 𝒢 ≤ ℋ) (hc : IsCusp c 𝒢) : IsCusp c ℋ :=
+  match hc with | ⟨h, hh, hp, hc⟩ => ⟨h, hGH hh, hp, hc⟩
+
+lemma IsCusp.of_relindex_ne_zero {c : OnePoint ℝ} (hGH : 𝒢.relIndex ℋ ≠ 0) (hc : IsCusp c ℋ) :
+    IsCusp c 𝒢 := by
+  rw [← Subgroup.inf_relIndex_right] at hGH
+  rw [← isCusp_iff_of_relIndex_ne_zero inf_le_right hGH] at hc
+  exact hc.mono inf_le_left
+
+open Pointwise in
+lemma Subgroup.conjAct_pointwise_smul_iff {G : Type} [Group G] {H : Subgroup G} {g : G} :
+    ConjAct.toConjAct g • H = H ↔ g ∈ normalizer H := by
+  rw [← H.normalizer.inv_mem_iff]
+  simp only [Subgroup.ext_iff, mem_pointwise_smul_iff_inv_smul_mem,
+    ← ConjAct.toConjAct_inv, ConjAct.toConjAct_smul, mem_normalizer_iff, inv_inv, Iff.comm]
+
+open Pointwise in
+lemma Subgroup.conjAct_pointwise_smul_eq_self
+    {G : Type} [Group G] {H : Subgroup G} {g : G} (hg : g ∈ normalizer H) :
+    ConjAct.toConjAct g • H = H :=
+  Subgroup.conjAct_pointwise_smul_iff.2 hg
+
+local notation "𝒬" => ℋ ⧸ (𝒢.subgroupOf ℋ)
+
+instance : MulAction ℋ ℋ := Monoid.toMulAction ..
+instance : MulAction ℋ 𝒬 := .quotient ..
+
+section SlashInvariantForm
+
+variable [SlashInvariantFormClass F 𝒢 k]
+
+/-- For `f` invariant under `𝒢`, this is a function on `(ℋ ⧸ 𝒢 ⊓ ℋ) × ℍ → ℂ` which packages up the
+translates of `f` by `ℋ`. -/
+def quotientFunc (q : 𝒬) (τ : ℍ) : ℂ :=
+  q.liftOn (fun g ↦ ((f : ℍ → ℂ) ∣[k] g.val⁻¹) τ) (fun h h' hhh' ↦ by
+    obtain ⟨j, hj, hj'⟩ : ∃ g ∈ 𝒢, h' = h * g := by
+      rw [← Quotient.eq_iff_equiv, Quotient.eq, QuotientGroup.leftRel_apply] at hhh'
+      exact ⟨h⁻¹ * h', hhh', mod_cast (mul_inv_cancel_left h h').symm⟩
+    simp [hj', SlashAction.slash_mul, SlashInvariantFormClass.slash_action_eq f j⁻¹ (inv_mem hj)])
```

Comment:
I found it surprising that we dont have some API already to do this. Should we not be doing this more generally, since it seems in lots of places we have the situation of taking some H-invariant object and turning it into a G-invariant one by using a norm or trace (for H < G some fin index subgroup).